### PR TITLE
LPD-7195 Send notifications to System Administrator when publication fails to publish due to java errors

### PR DESCRIPTION
**Pull request by gislayne-vitorino**
This is an update for: [LPD-7195](https://liferay.atlassian.net/browse/LPD-7195)

This is a look on the UI with the changes that add a new page to see the stacktrace
image

### CODE REVIEW

**Comment by gislayne-vitorino** [comment]
ci:test:publications

File: modules/apps/change-tracking/change-tracking-service/src/main/java/com/liferay/change/tracking/internal/background/task/CTPublishBackgroundTaskExecutor.java [hunk 1]
```diff
@@ -306,6 +338,22 @@ public void setAopProxy(Object aopProxy) {
...
+			companyId, RoleConstants.ADMINISTRATOR);
+
+		long[] userIds = _userLocalService.getRoleUserIds(role.getRoleId());
+
+		if (userIds.length == 0) {
+			throw new NoSuchUserException(
+				StringBundler.concat(
+					"No user exists in company ", companyId, " with role ",
+					role.getName()));
+		}
```
Comment (pei-jung) [inline modules/apps/change-tracking/change-tracking-service/src/main/java/com/liferay/change/tracking/internal/background/task/CTPublishBackgroundTaskExecutor.java:352]: This check is not needed because there will always be at least one administrator in the company.

File: modules/apps/change-tracking/change-tracking-service/src/main/java/com/liferay/change/tracking/internal/background/task/CTPublishBackgroundTaskExecutor.java [hunk 2]
```diff
@@ -283,13 +288,40 @@ public String handleException(
...
+							).put(
+								"isAdmin", true
+							).put(
+								"notificationType",
+								UserNotificationDefinition.
+									NOTIFICATION_TYPE_REVIEW_ENTRY
+							).put(
+								"showConflicts", showConflicts
+							));
+				}
```
Comment (pei-jung) [inline modules/apps/change-tracking/change-tracking-service/src/main/java/com/liferay/change/tracking/internal/background/task/CTPublishBackgroundTaskExecutor.java:324]: This logic is actually being added with Brooke's change in https://github.com/brianchandotcom/liferay-portal/pull/147360/commits/a31269bb97268b9e6f7e075f26b50e691f6223a4

File: modules/apps/change-tracking/change-tracking-web/src/main/java/com/liferay/change/tracking/web/internal/notifications/ScheduledPublicationUserNotificationHandler.java [hunk 3]
```diff
@@ -42,10 +42,28 @@ protected String getBody(
 		JSONObject jsonObject = _jsonFactory.createJSONObject(
 			userNotificationEvent.getPayload());
 
-		boolean showConflicts = jsonObject.getBoolean("showConflicts");
+		boolean admin = jsonObject.getBoolean("isAdmin");
```
Comment (pei-jung) [inline modules/apps/change-tracking/change-tracking-web/src/main/java/com/liferay/change/tracking/web/internal/notifications/ScheduledPublicationUserNotificationHandler.java:45]: We don't need a new flag variable. You can use `showConflicts` to determine the cause of the failure. If `showConflicts` is false, that means the failure is due to java exceptions.

File: modules/apps/portal-language/portal-language-lang/src/main/resources/content/Language.properties [hunk 4]
```diff
@@ -19503,6 +19503,7 @@ to-is-required=To is required.
 to-manage-priorities,-drag-the-rule-to-the-desired-position=To manage priorities, drag the rule to the desired position.
 to-navigate=to navigate
 to-save-this-form-all-required-fields-of-the-selected-object-need-to-be-mapped=To save this form, all required fields of the selected object need to be mapped.
+to-see-the-stack-trace-click-on-this-notification=To see the stack trace click on this notification.
```
Comment (pei-jung) [inline modules/apps/portal-language/portal-language-lang/src/main/resources/content/Language.properties:19506]: Reword this to be consistent with `click-on-this-notification-to-see-the-list-of-conflicts-that-need-to-be-manually-resolved`

File: modules/apps/portal-language/portal-language-lang/src/main/resources/content/Language.properties [hunk 5]
```diff
@@ -21402,6 +21403,7 @@ x-role={0} Role
 x-rows-were-imported-successfully={0} rows were imported successfully.
 x-rows-were-not-imported={0} rows were not imported.
 x-scheduled-publication-failed={0} scheduled publication failed.
+x-scheduled-publication-failed-with-java-error-not-related-to-conflicts={0} scheduled publication failed with java error not related to conflicts.
```
Comment (pei-jung) [inline modules/apps/portal-language/portal-language-lang/src/main/resources/content/Language.properties:21406]: Reword to `{0} scheduled publication failed with an unexpected system error.`

File: modules/apps/change-tracking/change-tracking-web/src/main/resources/META-INF/resources/publications/view_stack_trace.jsp [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+long backgroundTaskId = GetterUtil.getLong(renderRequest.getParameter("backgroundTaskId"));
+
+BackgroundTask backgroundTask = BackgroundTaskLocalServiceUtil.getBackgroundTask(backgroundTaskId);
+%>
+
+<clay:container-fluid>
+	<div>
+		<liferay-ui:message key="<%= backgroundTask.getStatusMessage() %>" />
+	</div>
+</clay:container-fluid>
```
Comment (pei-jung) [inline modules/apps/change-tracking/change-tracking-web/src/main/resources/META-INF/resources/publications/view_stack_trace.jsp:20]: Can you share a screenshot for what this looks like in the UI?

File: modules/apps/change-tracking/change-tracking-test/src/testIntegration/java/com/liferay/change/tracking/internal/test/PublicationUserNotificationHandlerTest.java [hunk 7]
```diff
@@ -306,6 +312,260 @@ public void testGetLinkToViewConflicts() throws Exception {
...
+		if (userIds.length == 0) {
+			throw new NoSuchUserException(
+				StringBundler.concat(
+					"No user exists in company ", companyId, " with role ",
+					role.getName()));
+		}
+
+		return userIds[0];
+	}
+
```
Comment (pei-jung) [inline modules/apps/change-tracking/change-tracking-test/src/testIntegration/java/com/liferay/change/tracking/internal/test/PublicationUserNotificationHandlerTest.java:571]: Consolidate the test cases so that we don't need to do the same test setup multiple times.

**Review by pei-jung — CHANGES_REQUESTED** [review]
@gislayne-vitorino please make sure to rebase your changes on top of https://github.com/brianchandotcom/liferay-portal/pull/147360. If it's not in upstream yet, rebase onto BChan's branch. Brooke refactored the `PublicationUserNotificationHandlerTest` quite a bit so it will conflict with your test changes.

File: modules/apps/change-tracking/change-tracking-web/src/main/java/com/liferay/change/tracking/web/internal/notifications/ScheduledPublicationUserNotificationHandler.java (hunk 3 above)
Comment (gislayne-vitorino) [reply modules/apps/change-tracking/change-tracking-web/src/main/java/com/liferay/change/tracking/web/internal/notifications/ScheduledPublicationUserNotificationHandler.java:45]: One of the notifications caused by the conflicts should be sent only to the system admin, so we need to check  that it is being sent directly to them

File: modules/apps/change-tracking/change-tracking-web/src/main/resources/META-INF/resources/publications/view_stack_trace.jsp (hunk 6 above)
Comment (gislayne-vitorino) [reply modules/apps/change-tracking/change-tracking-web/src/main/resources/META-INF/resources/publications/view_stack_trace.jsp:20]: This is what it looks on the UI with the changes  Pasted image

**Comment by gislayne-vitorino** [comment]
ci:test:sf

**Comment by gislayne-vitorino** [comment]
ci:test:publications

File: modules/apps/change-tracking/change-tracking-web/src/main/resources/META-INF/resources/publications/view_stack_trace.jsp (hunk 6 above)
Comment (pei-jung) [reply modules/apps/change-tracking/change-tracking-web/src/main/resources/META-INF/resources/publications/view_stack_trace.jsp:20]: Let's wrap the stack trace container in a sheet by adding these classes to the div `<div class="sheet sheet-lg">`. We will also need a heading to indicate what the stack trace is for. Maybe add this to the header: `P1 failed to publish.`

**Comment by gislayne-vitorino** [comment]
ci:test:sf

File: modules/apps/change-tracking/change-tracking-web/src/main/resources/META-INF/resources/publications/view_stack_trace.jsp (hunk 6 above)
Comment (gislayne-vitorino) [reply modules/apps/change-tracking/change-tracking-web/src/main/resources/META-INF/resources/publications/view_stack_trace.jsp]: After these changes: image

**Comment by pei-jung** [comment]
See https://github.com/liferay-publications/liferay-portal/pull/1052. Thanks!
